File: attempt1/go-attempt/src/network.go
package nnn

type Network struct {
	neurons []*Neuron
}

func NewNetwork() *Network {
	return &Network{[]*Neuron{}}
}

func (self *Network) AddNeuron(neuron *Neuron) {
	self.neurons = append(self.neurons, neuron)
}

func (self *Network) RemoveNeuron(neuron *Neuron) {
	for i, x := range self.neurons {
		if x == neuron {
			self.neurons[i] = self.neurons[len(self.neurons) - 1]
			self.neurons = self.neurons[0 : len(self.neurons) - 1]
			break
		}
	}
}

func (self *Network) Cycle() {
	for i := 0; i < len(self.neurons); i++ {
		self.neurons[i].willFire = self.neurons[i].NextCycle()
	}
	for i := 0; i < len(self.neurons); i++ {
		self.neurons[i].firing = self.neurons[i].willFire
	}
}

File: attempt1/failed/cpp-attempt/src/network.cpp
#include "network.hpp"
#include "link.hpp"

namespace nnn1 {
  
Network::~Network() {
  while (firstNeuron) {
    Neuron * next = firstNeuron->nextNeuron;
    delete firstNeuron;
    firstNeuron = next;
  }
}

void Network::AddNeuron(Neuron & neuron) {
  neuron.nextNeuron = firstNeuron;
  firstNeuron = &neuron;
  ++neuronCount;
}

void Network::RemoveNeuron(Neuron & neuron) {
  if (neuron.lastNeuron) {
    neuron.lastNeuron->nextNeuron = neuron.nextNeuron;
  } else {
    firstNeuron = neuron.nextNeuron;
  }
  if (neuron.nextNeuron) {
    neuron.nextNeuron->lastNeuron = neuron.lastNeuron;
  }
  --neuronCount;
}

void Network::Cycle() {
  ++cycleCount;
  
  // Record which neurons will fire on the next cycle
  Neuron * neuron = firstNeuron;
  while (neuron) {
    neuron->willFire = neuron->Cycle();
    neuron = neuron->nextNeuron;
  }
  // Fire away!
  neuron = firstNeuron;
  while (neuron) {
    if (neuron->willFire) {
      // The neuron is being fired during this cycle.
      neuron->GetLifeStatus().SetLastUsed(cycleCount);
    }
    if (neuron->isFiring) {
      // The neuron was fired during the last cycle, so all its outputs should
      // be updated accordingly.
      Link * link = neuron->firstOutput;
      while (link) {
        link->GetLifeStatus().SetLastUsed(cycleCount - 1);
        link = link->GetSenderNext();
      }
    }
    neuron->isFiring = neuron->willFire;
    neuron->willFire = false;
    neuron = neuron->nextNeuron;
  }
}

unsigned int Network::CountFiring() {
  unsigned int count = 0;
  Neuron * neuron = firstNeuron;
  while (neuron) {
    if (neuron->IsFiring()) {
      ++count;
    }
    neuron = neuron->nextNeuron;
  }
  return count;
}

void Network::UpdateLivesWithPain(double intensity) {
  Neuron * neuron = firstNeuron;
  while (neuron) {
    neuron->GetLifeStatus().UpdateNetPain(cycleCount, intensity);
    Link * link = neuron->firstOutput;
    while (link) {
      link->GetLifeStatus().UpdateNetPain(cycleCount, intensity);
      link = link->GetSenderNext();
    }
    neuron = neuron->nextNeuron;
  }
}

}
